backend-api/src/functions/adapters/aws/dynamo/dynamoDbAdapter.ts

```
@@ -131,6 +143,67 @@ export class DynamoDbAdapter implements SessionRegistry {
     }
   }
 
+  async getSession(
+    sessionId: string,
+    getOperation: GetSessionOperation,
+  ): Promise<Result<SessionAttributes, GetSessionFailed>> {
+    const getItemCommandInput = getOperation.getDynamoDbGetCommandInput({
+      tableName: this.tableName,
+      keyValue: sessionId,
+    });
+
+    let response;
+    try {
+      logger.debug(LogMessage.GET_SESSION_ATTEMPT, {
+        data: { sessionId, getItemCommandInput },
+      });
+
+      response = await this.dynamoDbClient.send(
+        new GetItemCommand(getItemCommandInput),
+      );
+    } catch (error: unknown) {
+      return this.handleGetSessionInternalServerError({
+        error,
+        getItemCommandInput,
+      });
+    }
+
+    const responseItem = response.Item;
+    if (responseItem == null) {
+      return this.handleGetSessionNotFoundError({
+        errorMessage: sessionNotFound,
+      });
+    }
+
+    // Attribute type validation
+    const getSessionAttributesResult =
+      getOperation.getSessionAttributesFromDynamoDbItem(responseItem);
+    if (getSessionAttributesResult.isError) {
```

So, if getting the attributes fails, we'll treat it as session not found (and down the line, the caller will get a 401, indicating that it's their fault). But if a session is corrupted/malformed in this way - e.g. one of its attributes is of the wrong type, or the session object is null, or something like that - doesn't that imply that something in _our_ system is wrong, rather than a client error? I wonder if we ought to be treating these failures as server errors (so long as we can still identify when it's a well-formed session in an incorrect state and return the right error). CC @sandymay-dd

---

Hopefully I haven't got the wrong end of the stick:

If we have type/structural issues with our session record, we should return a server error.

If we have a value validation failure (e.g. session age), we return a 401 Unauthorized. 

I thought this was the pattern we'd implemented for the update operations 👀

---

That's the right end of the stick, unless I've also got the wrong end of the stick. So, currently, it looks like we're returning a 401 unauthorized for type/structural issues (which are surfaced by `getSessionAttributesFromDynamoDbItem`. I propose that we instead return a server error in this scenario.